odoo/fields.py

```diff
@@ -2642,145 +2613,122 @@ def read(self, records):
+                cond=" AND ".join(clauses) if clauses else "1=1",
             )
-            cr.execute(query, tuple(links))
+            ids = {rid for recs, cs in records_commands_list for rid in recs.ids}
+            cr.execute(query, [tuple(ids)] + params)
+            for id1, id2 in cr.fetchall():
+                relation0[id1].add(id2)
+
+        # determine new relation {id1: ids2}
+        relation1 = defaultdict(set)
+        for x, ys in relation0.items():
+            relation1[x] = set(ys)
+
+        # operations on new relation
+        def relation_add(xs, y):
+            for x in xs:
+                relation1[x].add(y)
+
+        def relation_remove(xs, y):
+            for x in xs:
+                relation1[x].discard(y)
+
+        def relation_set(xs, ys):
+            for x in xs:
+                relation1[x] = set(ys)
+
+        def relation_delete(ys):
+            # the pairs (x, y) have been cascade-deleted from relation
+            for ys1 in relation0.values():
+                ys1.difference_update(ys)
+            for ys1 in relation1.values():
+                ys1.difference_update(ys)
+
+        to_create = []                  # line vals to create [(ids, vals)]
+        to_delete = []                  # line ids to delete
 
-        # remove links
-        links = old_links - new_links
-        if links:
-            cond = "{id1}=%s AND {id2}=%s".format(id1=self.column1, id2=self.column2)
-            query = """
-                DELETE FROM {rel} WHERE {cond}
-            """.format(
-                rel=self.relation,
-                cond=" OR ".join([cond] * len(links)),
+        with model.env.norecompute():
+            for records, commands in records_commands_list:
+                for act in (commands or ()):
+                    if not isinstance(act, (list, tuple)) or not act:
+                        continue
+                    if act[0] == 0:
+                        to_create.append((records._ids, act[2]))
+                    elif act[0] == 1:
+                        comodel.browse(act[1]).write(act[2])
+                    elif act[0] == 2:
+                        to_delete.append(act[1])
+                    elif act[0] == 3:
+                        relation_remove(records._ids, act[1])
+                    elif act[0] == 4:
+                        relation_add(records._ids, act[1])
+                    elif act[0] in (5, 6):
+                        # new lines must no longer be linked to records
+                        to_create = [(set(ids) - set(records._ids), vals)
+                                     for (ids, vals) in to_create]
+                        relation_set(records._ids, act[2] if act[0] == 6 else ())
+
+            if to_create:
+                # create lines in batch, and link them
+                lines = comodel.create([vals for ids, vals in to_create])
+                for line, (ids, vals) in pycompat.izip(lines, to_create):
```

Comment:
`pycompat` in master = :skull_and_crossbones:

Reply:
yeah, but `zip` is forbidden by pylint...